File: src_fortran/sub/reviveobs.f
c
c
c
c
c
c
      subroutine REVIVEOBS(i,nobs,iresflag)
c
c     If observationweight < 4 and weight=0.0 then set 'original' weight
c     if residual has become smaller since it has been reseted:
c
      implicit none
      integer i,nobs,iresflag
      include '../inc/vel_com.inc'
c
      real wsum,xkndw
      integer knobst,iii,iobswt
      if(w(nobs,i).ne.0.0) RETURN
      if(kpwt(nobs,i).eq.4) RETURN
c
cc         w(nobs,i)=0.0039  !  = 1./(2**(2*reading_weight)) = 1./256
c
c     re-normalize weights of this event :
c
      nobswithw0=nobswithw0-1
      wsum=0.0
      knobst=knobs(i)
      do iii=1,knobst
         iobswt=kpwt(iii,i)
         if(iii.ne.nobs.and.w(iii,i).eq.0.0)then
            w(iii,i)=0.0   ! keep zero-weights as they are !!!
         else
            if(iobswt.lt.4)then
               if(sphase(iii,i).eq.1.0.or.sphase(iii,i).eq.2.0)then
                  w(iii,i)=swtfac*1.0/(2**(iobswt*2))   ! S-phase
               else
                  w(iii,i)=       1.0/(2**(iobswt*2))   ! P- or M-phase
               endif
            else
               w(iii,i)=0.0   ! observation-weight 4 ==> don't use this arrival
            endif
         endif
         wsum=wsum+w(iii,i)
      enddo
      xkndw=float(knobst-nobswithw0)/wsum
      do iii=1,knobst
         w(iii,i)=w(iii,i)*xkndw
      enddo
      if(.not.single_turbo)then
      write(16,*)'WARNING:'
      write(16,'('' Iteration# '',i3,'' Event#'',i4,'' Res#'',i3,
     &           '' ='',f7.2,''; ABS < 1.0 ---> weight revived !'')')
     &           nitt,isingle,nobs,res(nobs,i)
      endif
      write(6,'('' Iteration# '',i3,'' Event#'',i4,'' Res#'',i3,
     &           '' ='',f7.2,''; ABS < 1.0 ---> weight revived !'')')
     &           nitt,isingle,nobs,res(nobs,i)
      if(.not.single_turbo)then
         write(16,*)'knobs(i)   = ',knobst
         write(16,*)'nobswithw0 = ',nobswithw0
      endif
cc      write(6,*)'knobs(i)   = ',knobst
cc      write(6,*)'nobswithw0 = ',nobswithw0
cc      write(6,*)
      iresflag=1  ! inhibit a pre-stopping in subr. OUTPUT due to
                  ! increased datvar; give location a second chance.
c
      return
      end ! of subr. reviveobs

File: src/sub/reviveobs.c
/*
 * Revive observation with small residual that was previously rejected
 * Re-normalize weights after revival
 * from Fortran REVIVEOBS subroutine
 */

#include <stddef.h>
#include <stdio.h>
#include <stdbool.h> 

/* External global variables from vel_com.inc */
extern int nitt;
extern int isingle;
extern int nobswithw0;
extern bool single_turbo;
extern double w[][100];
extern int knobs[];
extern int kpwt[][100];
extern double sphase[][100];
extern double swtfac;
extern double res[][100];

void reviveobs(int i, int nobs, int *iresflag) {
    double wsum, xkndw;
    int knobst, iii, iobswt;
    
    /* Check if weight is not zero (already active) */
    if (w[nobs - 1][i] != 0.0) {
        return;
    }
    
    /* Check if observation weight is 4 (don't use) */
    if (kpwt[nobs - 1][i] == 4) {
        return;
    }
    
    /* Re-normalize weights for this event */
    nobswithw0--;
    wsum = 0.0;
    knobst = knobs[i];
    
    for (iii = 0; iii < knobst; iii++) {
        iobswt = kpwt[iii][i];
        
        /* Keep zero weights as they are, except for current observation */
        if (iii != (nobs - 1) && w[iii][i] == 0.0) {
            w[iii][i] = 0.0;
        } else {
            /* Assign weight based on observation class */
            if (iobswt < 4) {
                if (sphase[iii][i] == 1.0f || sphase[iii][i] == 2.0f) {
                    w[iii][i] = swtfac * 1.0f / (double)(1 << (iobswt * 2));  /* S-phase */
                } else {
                    w[iii][i] = 1.0f / (double)(1 << (iobswt * 2));  /* P or M phase */
                }
            } else {
                w[iii][i] = 0.0;  /* Weight 4 ==> don't use */
            }
        }
        wsum += w[iii][i];
    }
    
    /* Normalize weights */
    xkndw = (double)(knobst - nobswithw0) / wsum;
    for (iii = 0; iii < knobst; iii++) {
        w[iii][i] = w[iii][i] * xkndw;
    }
    
    /* Output warning messages */
    if (!single_turbo) {
        fprintf(stderr, "WARNING:\n");
        fprintf(stderr, " Iteration# %3d Event#%4d Res#%3d =%.2f; "
                "ABS < 1.0 ---> weight revived !\n",
                nitt, isingle, nobs, res[nobs - 1][i]);
    }
    printf(" Iteration# %3d Event#%4d Res#%3d =%.2f; "
           "ABS < 1.0 ---> weight revived !\n",
           nitt, isingle, nobs, res[nobs - 1][i]);
    
    if (!single_turbo) {
        fprintf(stderr, "knobs(i)   = %d\n", knobst);
        fprintf(stderr, "nobswithw0 = %d\n", nobswithw0);
    }
    
    /* Signal to inhibit pre-stopping in OUTPUT */
    *iresflag = 1;
}
